share/steel/examples/pulse/lib/pledge/Pulse.Lib.Trade.fst

```diff
@@ -0,0 +1,143 @@
+(*
+   Copyright 2023 Microsoft Research
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*)
+
+module Pulse.Lib.Trade
+
+open Pulse.Lib.Core
+open Pulse.Lib.Pervasives
+open Pulse.Lib.InvList
+module T = FStar.Tactics
+
+let trade_elim_t is hyp extra concl : Type u#2 =
+  unit -> stt_ghost unit emp_inames (invlist_v is ** extra ** hyp) (fun _ -> invlist_v is ** concl)
+
+let trade_elim_exists (is:invlist) (hyp extra concl : vprop) : vprop =
+  pure (squash (trade_elim_t is hyp extra concl))
+
+let trade (#is : invlist) (hyp : vprop) (concl : vprop) =
+  exists* extra. extra ** trade_elim_exists is hyp extra concl
+
+```pulse
+ghost
+fn __intro_trade
+  (#is : invlist)
+  (hyp concl : vprop)
+  (extra : vprop)
+  (f_elim: unit -> (
+    stt_ghost unit emp_inames
+    (invlist_v is ** extra ** hyp)
+    (fun _ -> invlist_v is ** concl)
+  ))
+  requires extra
+  ensures trade #is hyp concl
+{
+  fold (trade_elim_exists is hyp extra concl);
+  assert (extra ** trade_elim_exists is hyp extra concl); // FIXME: why is this needed? somehow guiding the prover?
```

Comment:
I think this is because the context is:

```
trade_elim_exists is hyp extra concl ** extra
```

The fold makes the goal `exists* extra. extra ** trade_elim_exists is hyp extra concl`, which after an intro exists step becaomes `?u ** trade_elim_exists is hyp ?u concl`. The unifier is greedy, left-to-right, and does not back track. So, this solves `?u := trade_eilm_exists is hyp extra concl` and then we're left with trying to match `extra` with `trade_elim_exists ...` which fails. The assert rewrites the context to match the left-to-right order of the goal. Another fix would be to flip the order of conjuncts in the definition of `trade`, so that the least constrained sub-term (`extra`) is last, i.e., right most.

In any case, this order sensitivity is a unfortunate, but without a bit of backtracking, it seems unavoidable? Or perhaps we can try to find and solve rigid-rigid matches first, before trying any flex-rigid case.